### PR TITLE
HTML API: Allow additional fragment contexts.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac-ticket: Core-61576.

## Status

 - [ ] ~Add unit tests covering significant contexts, e.g. SCRIPT, TEXTAREA, SVG.~
    - It shouldn't be possible to create fragment parsers on void or self-contained elements. There's no content inside of them to parse. For self-contained elements it's appropriate to rely on the eventual `set_inner_html()` which will be an alias in those cases for `set_inner_text()`.
 - [ ] Review all documentation looking for places stating that `<body>` is the only supported context.

## Description

Previously, the fragment parser in WP_HTML_Processor has only allowed creating a fragment with the `<body>` context. In this patch, any context node is allowed.

### CODE REVIEW

**Comment by dmsnell** [comment]
@sirreal it looks like we may have a lot of work to do before we can open this up to allowing self-contained elements as the fragment context. there are multiple challenges:

 - we have to enter within a certain tokenizer state, but that's not how our parser is designed.
 - the DOM is able to store things like the text `</textarea>` inside a `TEXTAREA` node, but HTML is not able to do this. we have to determine what to do in these cases.
 - if the closing tag for the self-contained element isn't provided, we cannot parse it to figure out which context node it is. e.g. `<script>` bails with the tag processor.

**Comment by sirreal** [comment]
The fragment parser takes an Element. An HTML string seems like a poor representation for that 🙂 

I wonder if the fragment parser could take a token (to create a fragment from an existing document) or some other representation that's not just a string in order to create fragments without needing to create a full document first.

> The following steps form the HTML fragment parsing algorithm. The algorithm takes as input an [Element](https://dom.spec.whatwg.org/#interface-element) node, referred to as the context element, which gives the context for the parser, input, a string to parse, and an optional boolean allowDeclarativeShadowRoots (default false). It returns a list of zero or more nodes.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -281,24 +281,41 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
...
 	 * @param string $encoding Text encoding of the document; must be default of 'UTF-8'.
 	 * @return static|null The created processor if successful, otherwise null.
 	 */
 	public static function create_fragment( $html, $context = '<body>', $encoding = 'UTF-8' ) {
-		if ( '<body>' !== $context || 'UTF-8' !== $encoding ) {
+		if ( 'UTF-8' !== $encoding ) {
+			return null;
+		}
+
+		$context_processor = new WP_HTML_Tag_Processor( $context );
```
Comment (apermo) [inline src/wp-includes/html-api/class-wp-html-processor.php]: Not sure how deep you want to comment, but a comment here could help understanding what you were up to.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -281,24 +281,41 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
...
+		if ( ! $context_processor->next_token() || '#tag' !== $context_processor->get_token_type() ) {
+			return null;
+		}
+
+		$context_tag        = $context_processor->get_tag();
+		$context_attributes = array();
+		foreach ( $context_processor->get_attribute_names_with_prefix( '' ) as $name ) {
```
Comment (apermo) [inline src/wp-includes/html-api/class-wp-html-processor.php]: Having an empty prefix here looks strange.
If it was my code I'd either have the prefix optional/default to empty or add a `get_attributes_names()` which would do exactly the same as `get_attributes_names_with_prefix( '' )`, might be redundant in a way, but for me it would look a little more readable.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 3]
```diff
@@ -281,24 +281,41 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
...
 	 *
+	 * @todo Verify that creating a fragment in self-contained elements works.
+	 *
 	 * @since 6.4.0
 	 * @since 6.6.0 Returns `static` instead of `self` so it can create subclass instances.
+	 * @since 6.7.0 Can create fragment in any context.
 	 *
 	 * @param string $html     Input HTML fragment to process.
 	 * @param string $context  Context element for the fragment, must be default of `<body>`.
 	 * @param string $encoding Text encoding of the document; must be default of 'UTF-8'.
```
Comment (apermo) [inline src/wp-includes/html-api/class-wp-html-processor.php]: "must be default of ..."? 
My internal autocomplete expected an "or" here. Which likely is equivalent, but I was expecting it.

And I think at least for `$context` you need to update it, since your change was about allowing other than body.

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 4]
```diff
@@ -281,24 +281,41 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
...
+		}
+
+		if ( $context_processor->next_token() ) {
 			return null;
 		}
 
 		$processor                             = new static( $html, self::CONSTRUCTOR_UNLOCK_CODE );
-		$processor->state->context_node        = array( 'BODY', array() );
+		$processor->state->context_node        = array( $context_tag, $context_attributes );
 		$processor->state->insertion_mode      = WP_HTML_Processor_State::INSERTION_MODE_IN_BODY;
```
Comment (apermo) [inline src/wp-includes/html-api/class-wp-html-processor.php:340]: If I understood your PR correctly, it's about allowing insertion to anything other than body, or is body in this case ambivalent?

So content body vs `<body>`?

Anyways, I'm uncertain wether this is intentional or if you forgot to touch this.

**Review by apermo — COMMENTED** [review]
As you asked me, here are my thoughts.

One additional one. Do you want to add a blacklist of forbidden subnodes? Or do you think this is the responsibility of the developer?

Like preventing to add `<html>`, `<body>` to `body`, or pretty much everything to `title`.

Not sure wether this would be the right place to do so, but I had that thought and wanted to share it.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php]: thanks @apermo. if this looks strange, it's supposed to 🙃 
when we added this method we wanted it to be a bit awkward in the hopes of communicating that it involves additional costs. we wanted the default behavior to be looking for a subset of attributes (such as `get_attribute_names_with_prefix( 'data-' )`) so that the API itself guides people to learn in the safest most performant manner.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 3 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php]: thanks for noticing.

`must be default of UTF-8` and `must be default or UTF-8` are very different, whereas only the default value has been allowed (and the default is UTF-8). some day we might open it up to other values, but this is there to communicate intentionally that this is a UTF-8-only interface at the moment.

File: src/wp-includes/html-api/class-wp-html-processor.php (hunk 4 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:340]: the fragment parser is synonymous with `node.innerHTML = newHtml` in JavaScript or with a DOM. in this case, `<body>` has been the default as a reasonably safe default for existing code, which likely is getting only a small chunk of the actual site HTML and then trying to process it.

so if you knew you were inside a `<li>` you would create the fragment parser in the `<li>` context and then the parse would change based on that. I probably don't fully understand this, because it's hard for me to come up with situations that lead to different parses, but it comes into play when inside SVGs or MathML elements, and when resetting some internals (the insertion mode).

basically this is not something most people will need to use, but it will be used by `set_inner_html()` to ensure appropriate parsing.

**Comment by dmsnell** [comment]
@sirreal: upon further review I feel like this is not appropriate for self-contained and void elements, so I am rejecting them. there's never a reason, I think, to create a fragment parser on an element which can contain no child nodes (other than text, potentially).

One case I was specifically thinking through is a `<script>` context. The idea is simple: parse the inner HTML of a SCRIPT element. The problem is when `</script>` appears and there is more content. In the DOM you can actually create a text node containing `</script>` but this is not possible in the HTML. The spec even notes situations like this where the DOM can create trees that cannot be expressed in HTML.

So I think that long-term the only use for the fragment parser outside of the usual is when setting inner HTML, and in that case, it seems more fitting for _that_ function to call out something special for elements like `SCRIPT` so that it falls into the equivalent of `->skip_script_data()`.

Apart from this I don't like exposing a token as the context node because that's cumbersome for people to write, and I don't expect people to want to go through the hassle of what that implies. Our tokens don't even contain attribute references, so this is not currently workable.

However, it does potentially solve some issues around namespacing. There's no way in this function's argument to specify a context node in the `svg` or `math` namespace.

**Comment by sirreal** [comment]
I'll share some context from a conversation.

---

[As mentioned](https://github.com/WordPress/wordpress-develop/pull/7141#issuecomment-2300365464), it likely doesn't make sense to create a fragment parser in order to modify these "self-contained" tags and better interfaces can be provided.

---

The specification for fragment parsers are always created from another document. This should be easy to do in the HTML API. However, most of the time WordPress is inspecting or modifying HTML snippets without considering their context. Rarely are full HTML documents considered. In order to continue supporting this behavior, we need to be able to create fragments without a parent document. This is where most of the difficulty arises. Most of the time, `<body>` provides a suitable context, but there could be other cases like `<head>`. Again, this seems relatively straightforward to support. However, it gets more complicated with things like `<svg>` as a context element, or especially `<rect>`, or even trickier a script tag in the SVG namespace!

The context element will be used to establish the correct insertion mode _and_ the context element's namespace and possibly attributes determine how tokens should be handled (using rules for foreign content or one of the HTML insertion modes).

It may make sense to maintain this interface as-is. `<body>` is likely a good context for most use-cases but in order to properly support fragments some advanced interface likely needs to support:

- Passing a context element along with its tag name, namespace, and attributes.
- The context element's document determines the quirks mode of the fragment parser.

**Comment by sirreal** [comment]
> in order to properly support fragments some advanced interface likely needs to support…

I had a thought that may be helpful.

I'd like to add an instance method to the HTML processor like `create_fragment_parser_at_node( string $html_fragment )`. This method would return a new HTML processor, created as a fragment parser with the context element created from the current tag opener (or null). This would allow use to set the document compat type and the context element and its attributes correctly according to the specification.

Once we have this method, advanced fragment parser creation could be handled by that method:

```php
$full_parser = WP_HTML_Processor::create_full_parser( '<!DOCTYPE html><math><annotation-xml encoding="text/html">' );
$full_parser->next_tag( 'ANNOTATION-XML' );
$fragment_parser = $full_parser->create_fragment_parser_at_node( `<h1>Who knows what happens here?' );
$fragment_parser->next_tag( 'H1' );
// …
```

I think internally this method would be used for things like set_inner_html that require a fragment parser.

**Comment by sirreal** [comment]
See #7348

**Comment by sirreal** [comment]
https://github.com/WordPress/wordpress-develop/pull/7777 (on top of https://github.com/WordPress/wordpress-develop/pull/7348) adds the ability for advanced fragments, for example `WP_Html_Processor::create_fragment( '<rect/>', '<svg>' )` will do what you expect 🙂

**Comment by sirreal** [comment]
https://github.com/WordPress/wordpress-develop/pull/7777 is open for review. That would supersede this work.

**Comment by sirreal** [comment]
https://core.trac.wordpress.org/changeset/59467 / https://github.com/WordPress/wordpress-develop/pull/7777 have landed.

I think this PR can be closed.
